### ISSUE TITLE
qwen35: migrate GDN prefill to FlashInfer and retire Triton AOT where supported

### ISSUE THREAD

**Issue by xiaguan** [xiaguan/pegainfer#691]
## Context

OpenInfer pins FlashInfer `v0.6.14` at `19f1a41e6b21f0c422d775e377b6fdf9a1fc9d23`.

Qwen3.5 prefill currently runs a seven-stage Triton-AOT Gated Delta Rule pipeline generated from `openinfer-kernels/tools/triton/gated_delta_rule_chunkwise_kernels.py`. This is the Qwen3.5 model line's only build-time Triton dependency.

The pinned FlashInfer tree now has a fused `chunk_gated_delta_rule` prefill backend for the architectures relevant to current Qwen3.5 work:

- SM100: [flashinfer#3001](https://github.com/flashinfer-ai/flashinfer/pull/3001)
- SM90: [flashinfer#3477](https://github.com/flashinfer-ai/flashinfer/pull/3477)
- SM120: [flashinfer#3479](https://github.com/flashinfer-ai/flashinfer/pull/3479)
- Follow-up prefill work: [flashinfer#3742](https://github.com/flashinfer-ai/flashinfer/pull/3742)

FlashInfer's own benchmark matrix includes Qwen3.5 4B/9B (`Hq=Hk=16`, `Hv=32`), 27B (`Hv=48`), and larger-family head configurations. Its tests cover the 16/16/32 GVA shape used by 4B and 9B.

This makes an upstream-backed migration plausible. It is not a drop-in wrapper.

### 1. The pinned implementation is CuTe DSL, not a reusable C++ launcher

All pinned SM90/SM100/SM120 prefill backends live under `flashinfer/gdn_kernels/` and compile through Python CuTe DSL. GDN is not registered in `flashinfer/aot.py`, and the submodule contains no GDN cubin/fatbin/PTX artifact.

A direct adoption removes Triton but substitutes `nvidia-cutlass-dsl` plus Python artifact generation. Removing both Triton and Python requires one of:

- a reproducible generator plus checked-in or release-packaged per-architecture cubins;
- future upstream `flashinfer-cubin` coverage for GDN;
- another explicit AOT path that exposes a raw-pointer launch contract.

Do not add a Python runtime dependency to the Rust server.

### 2. OpenInfer still needs a small preparation kernel

The current first Triton stage splits fused QKV, expands and L2-normalizes Q/K, and computes the forget/update gates. FlashInfer supports Qwen3.5's GVA mapping directly, so Q/K expansion should be unnecessary, but a native CUDA preparation step is still needed to:

- split the fused projection into Q/K/V views or buffers;
- L2-normalize Q/K;
- convert Qwen3.5's log-space gate to FlashInfer's linear-space forget factor;
- produce fp32 gate and beta tensors.

### 3. Persistent state layout differs

FlashInfer documents final state as `[N,H,V,K]`. OpenInfer's native decode kernel documents and indexes `[H,K,V]`. Since `K=V=128`, shape checks cannot detect a semantic transpose.

Choose one persistent state schema and make prefill plus decode agree. Do not insert a per-layer steady-state transpose. Keeping the native decode kernel with `[H,V,K]` indexing is a valid first scope; migrating decode to FlashInfer is not required unless measurements justify it.

### 4. Architecture support is narrower

FlashInfer GDN prefill supports SM90/SM100/SM120. OpenInfer has historical Qwen3.5 correctness evidence on SM80.

The implementation must make one explicit decision:

- keep the Triton backend as an SM80 fallback;
- provide a non-Triton SM80 fallback;
- or raise the documented Qwen3.5 hardware floor.

With an SM80 Triton fallback, Qwen3.5 can stop using Triton on SM90+ but cannot remove the dependency globally.

## Proposed implementation sequence

1. Build a focused raw-pointer FlashInfer GDN prefill wrapper for one retained architecture first. SM120 is the current Qwen3.5 serving-performance anchor.
2. Add the native CUDA preparation kernel and feed FlashInfer unexpanded GVA Q/K/V.
3. Resolve the persistent state layout without a runtime transpose.
4. Add explicit architecture dispatch and fail closed when the selected artifact/backend is unavailable.
5. Compare peak scratch memory against `GdrChunkwiseScratch35`; delete old scratch only after the new path is proven.
6. Expand to SM90 and SM100, or document why an architecture remains on the existing backend.
7. Remove Triton build code and `OPENINFER_TRITON_PYTHON` only after no supported target depends on it.

## Correctness gates

- Focused output and final-state parity for sequence lengths spanning partial and full 64-token chunks.
- Head configurations at least 16/16/32 and 16/16/48.
- Prefill-to-decode handoff parity with a test that detects the K/V state transpose.
- Qwen3.5 4B, 9B, and 27B short and long HF logits gates.
- Scheduler-level chunked prefill, including resumed `base_pos > 0` execution.
- Existing decode CUDA Graph capture remains valid if the native decode state layout changes.
- Unsupported architectures and missing/mismatched artifacts fail at startup instead of silently falling back.

## Performance and build evidence

No OpenInfer A/B has been measured yet, so this issue makes no performance claim.

Before calling the migration a win, report same-host current-vs-FlashInfer measurements for:

- focused GDN prefill latency across representative prompt lengths;
- end-to-end Qwen3.5 TTFT with the same model, CUDA version, and workload;
- peak GDN scratch memory;
- build time and generated artifact size;
- any decode TPOT change caused by a persistent-state layout rewrite.

Keep upstream-reported improvements separate from OpenInfer measurements.

## Done criteria

- The selected FlashInfer SHA and generated-artifact provenance are recorded.
- The Rust/CUDA runtime has no Python dependency.
- The SM80/SM90/SM100/SM120 support matrix and fallback policy are explicit.
- Correctness gates above pass on every enabled backend.
- Measured A/B data supports keeping the new backend.
- `kernel_plan()`, Qwen3.5 build documentation, and feature descriptions match the final backend.
- Triton files, feature wiring, and environment handling are removed only if the final support matrix no longer needs them.

Related: #479 documented the general FlashInfer bump/JIT boundary; #517 tracked the old value-head specialization. Neither owns this migration.

**Comment by xiaguan** [comment]
Cross-reference with #690 (in flight), which touches this from the decode side:

- **Constraint 3 gets one more consumer.** #690 replaces the per-slot decode GDR/conv1d launches with batched kernels that index recurrent state through per-layer device pointer tables — and the new `gated_delta_rule_decode_batch_kernel` bakes the native `[H,K,V]` indexing into a second production kernel (the single-slot one is retained as a test reference). Once it lands, "make prefill and decode agree" means updating the batch kernel, and the prefill-to-decode handoff parity test proposed here is the thing that would catch a silent K/V transpose there too — worth writing it against the batched path, since that's what production runs.
- **The pointer-table mechanism is actually favorable for this migration.** Slot state buffers are opaque per-request allocations reached via a `u64` table; a persistent-layout change stays contained in kernel indexing, with no CUDA-graph re-plumbing — the graph only depends on buffer addresses, which don't move. The "decode CUDA Graph capture remains valid" gate gets easier, not harder.
- **Scoping evidence.** #690's measured decode win (HTTP c16 TPOT 17.35ms → 11.30ms on the retained SM120 host) strengthens the call in Constraint 3 that migrating decode to FlashInfer is not required for now — prefill-only scope with the native decode kernels kept looks right, and any future decode migration would need to beat a much better baseline.

Practical ordering note: if #690 merges first (likely), step 3 here should treat the batched decode kernel + pointer tables as the canonical decode-side state contract, not the retired per-slot kernel.

**Comment by Nyvo-io** [comment]
Hi @xiaguan, I would like to take the SM120-first slice from the proposed implementation sequence. I have an RTX 5090/SM120 validation host. My intended scope is: coordinate with #690 as the canonical batched-decode state contract; build a reproducible FlashInfer CuTe artifact path and raw-pointer Rust/CUDA wrapper with no Python runtime dependency; add the native Q/K normalization + gate preparation seam; and gate output/final-state plus prefill-to-decode parity across partial/full 64-token chunks and resumed `base_pos > 0`. I will also report focused latency, scratch-memory, and same-host TTFT A/B. I will not claim SM80/90/100 coverage or remove the Triton fallback until those backends and the support policy are proven. Please assign this incremental scope to me if it matches how you want #691 staged.

**Comment by Nyvo-io** [comment]
Following up to make ownership explicit: I am taking the SM120/RTX 5090 first-stage implementation for #691. I will coordinate the recurrent-state contract with #690, keep the SM80 Triton fallback intact, and submit measured correctness, scratch-memory, and TTFT evidence before proposing any broader architecture cleanup.

**Comment by Nyvo-io** [comment]
Status update on the SM120-first work for #691:

I opened PR #709 for the initial FlashInfer CuTe-DSL AOT integration, but closed it after the mandatory correctness gate failed. The same-GPU Triton baseline passed, while the FlashInfer path produced different tokens and failed the Qwen3.5 HF golden logits tolerance.

A local diagnostic fix corrected the generated TMA views for k/v/output and reduced the 32-value-head operator error substantially, but the full model-level gate still failed. I therefore stopped before TTFT/TPOT benchmarking and did not retain the second backend.

I would like to continue with a narrower correctness-first follow-up:

- preserve the 32-head operator regression;
- compare Triton and FlashInfer at partial/full 64-token boundaries;
- trace per-layer output and recurrent-state divergence;
- verify gate semantics, state layout, and prefill-to-decode handoff;
- only reopen an implementation PR after the HF golden gates pass.

Would you prefer this diagnostic work as a small standalone test PR first, or should I keep it local until the remaining model-level divergence is resolved?

**Comment by Nyvo-io** [comment]
Follow-up to my previous #691 update: I continued the diagnostic work locally after closing PR #709. The earlier HF-golden failure (`mean=0.6768`, `p99=2.3126`) was not the final kernel result; it exposed a second AOT issue after the initial TMA-view fix.

Root cause: the generator used a partially static fake-tensor layout, which produced a specialization not equivalent to FlashInfer's official DLPack dynamic layout. I changed the formal generator to one fully-dynamic fake-tensor path (dynamic token/head dimensions, non-leading strides, and auxiliary-buffer lengths) and kept the C shim descriptors explicit: q `[T,D,H]`, k/v/output `[D,T,H]`. The diagnostic environment-variable branches were removed.

The follow-up gates now pass on RTX 5090/SM120:
- operator boundary cases (`seq_len` 1/2/63/64/65/127/128, non-zero initial state, 32 value heads): max output diff about `0.000122`, max state diff about `0.000238`;
- HF golden: sequential mean/p99 `0.0272/0.0975`, batched-graph mean/p99 `0.0263/0.0953`, slot-compaction mean/p99 `0.0254/0.1104`, long-golden mean/p99 `0.0153/0.0617`, all within repository limits.

I then ran a correctness-gated same-host request A/B (prompt 128/1024/4096, concurrency 1/8/16, output 64). FlashInfer was not a stable performance winner: only prompt 4096/concurrency 16 improved TPOT by about 2.66%; several low/medium-load cells regressed, and a few high-concurrency cells had different token hashes. The scheduler-dependent `decode` harness is currently invalid for Qwen3.5 because `Scheduled.cached_tokens` is not provided, so the measurements use the request harness for TTFT/steady TPOT.

Recommendation: keep #709 closed for now. I can split the fully-dynamic AOT correctness/layout regression into a small reviewable test PR if that is useful, or continue tracing the remaining high-concurrency hash differences. I would not reopen the original opt-in dual-backend PR without a maintainer decision on the performance bar and single-backend policy.

**Comment by qwzx-qwas** [comment]
Hi @xiaguan  ，I’m very interested in this issue. I noticed that there has not been any new activity on it for the past two weeks, so I would like to take a shot at it.

For #691, I would like to work on the SM120/Qwen3.5-4B stage first. The problem I intend to address is replacing the current seven-stage Triton GDN prefill path without changing OpenInfer’s existing [H,K,V] persistent-state contract or introducing per-layer state transposes.

I plan to approach this by generating a reproducible, pinned FlashInfer CuTe artifact with a raw-pointer Rust/CUDA interface; adding a native CUDA preparation kernel that splits fused Q/K/V, L2-normalizes the original unexpanded Q/K heads, and produces the per-token alpha/beta expected by FlashInfer; and applying a narrow SM120 state-layout specialization so the FlashInfer kernel can read and write OpenInfer’s [H,K,V] state directly. The exported CUDA module will be owned by the corresponding model/device context rather than shared globally.

I will first validate the new path against the existing Triton prefill and native stepwise decode, including non-zero initial state, partial and full 64-token tiles, resumed chunked prefill, and the first decode step after prefill. Only after the Qwen3.5-4B short and long HF logits gates pass will I measure focused GDN latency, peak scratch memory, and same-host Triton-vs-FlashInfer TTFT/TPOT on an RTX 5090.

My initial production target is the Hq=16, Hk=16, Hv=32, D=128 SM120 configuration used by Qwen3.5-4B. I can add operator-level Hv=48 coverage to detect accidental 32-head specialization, but I will not claim full 27B, SM80/SM90/SM100, or multi-device support in this stage. I will also keep the Triton fallback until the remaining architecture matrix and artifact-distribution policy are proven.

If this incremental scope matches how you want #691 staged after #709, please assign it to me.

**Comment by qwzx-qwas** [comment]
I have now completed the operator-level correctness testing on an RTX 5090, covering two geometries (`Hq/Hk/Hv/D = 16/16/32/128` and `16/16/48/128`) across `T = 1, 2, 63, 64, 65, 127, 128`.

The Hv32 production-candidate geometry used by Qwen3.5-4B passes all cases. For the Hv48 generalization coverage, only `T=128` currently blocks the gate: the FlashInfer final state has `5/786432` localized tail violations relative to the FP64 oracle, or about `6.36 ppm`, while the output still passes.

To rule out any effect from the OpenInfer HKV state-layout patch, I ran an A/B comparison between the patched HKV artifact and an unmodified upstream HVK artifact using identical inputs. After converting the state layout appropriately, the two artifacts produced bitwise-identical outputs and final states. The coordinates and errors of all five violations were also identical. This confirms that the discrepancy is not introduced by the HKV patch, TMA indexing, or the alias path; it is already present in the upstream FlashInfer SM120 kernel.

The block-boundary diagnostics further show that `T128` is bitwise identical to `T64 + T64`, but differs from `T127 + T1`, while the standalone `T1` path is clean. My current interpretation is therefore that the tail error comes from rounding differences in the 64-token blockwise BF16/WGMMA numerical path accumulating through the recurrence.

Compared with the Triton baseline, FlashInfer performs better for most dynamic sequence lengths and also has better `max_abs`, `mean_abs`, and `p99_abs` at `T=128`. However, its localized tail at `T=128` is worse: FlashInfer has 5 violations, while Triton has 2.

At this point, I am unsure whether the next step should be to continue pursuing an FP32 correction specifically for the Hv48 path, or to treat this bounded upstream numerical tail as acceptable for the non-production generalization geometry.

**Comment by qwzx-qwas** [comment]
<html>
<body>
<p class="isSelectedEnd"><span>I have now completed the model-level HF accuracy gates and the same-machine RTX 5090 performance comparison.</span></p><p class="isSelectedEnd"><span>For the production Qwen3.5-4B Hv32 geometry, the FlashInfer candidate passes all operator-length tests, the short- and long-sequence HF fixtures, resumed prefill, the prefill-to-decode handoff, scheduler/slot-reuse tests, and CUDA Graph tests. The launch-count evidence also confirms that none of these tests silently fell back to Triton.</span></p><p class="isSelectedEnd"><span>Against the same HF oracle, both backends pass all five replay surfaces. FlashInfer has a lower mean error in 4/5 surfaces and a lower maximum error in 4/5, with an approximate weighted-mean improvement of 2.8%. However, it has a lower p99 in only 2/5 surfaces. Therefore, the appropriate conclusion is that FlashInfer is slightly better overall, but it is not uniformly better than Triton across every tail metric.</span></p><p class="isSelectedEnd"><span>On performance, for the tested c1/c4 cases, FlashInfer improves TTFT p50 by 1.79%–5.47% and throughput by 0.39%–2.09%. At c8, throughput improves by 1.86% and TTFT p99 improves by 3.83%, while TTFT p50 regresses by 4.00% and shows relatively high variance. Nsight shows that direct GDN kernel time is reduced by 5.93%, the number of GDN kernels per layer is reduced from seven to two, and scratch memory drops from 184.50 MiB to 48.52 MiB, a 73.70% reduction. The main cost is an approximately two-second one-time cold PTX JIT overhead.</span></p><p class="isSelectedEnd"><span>The detailed HF results are shown below; lower values are better:</span></p>
Replay surface | Triton mean / p99 / max | FlashInfer mean / p99 / max
-- | -- | --
Short sequential | 0.0268 / 0.0985 / 0.1827 | 0.0256 / 0.0992 / 0.1367
Batched, 5 padded | 0.0268 / 0.0973 / 0.1835 | 0.0273 / 0.1156 / 0.1781
Batched, 3 padded | 0.0285 / 0.1019 / 0.1850 | 0.0280 / 0.0878 / 0.1124
Slot compaction | 0.0267 / 0.1016 / 0.1829 | 0.0256 / 0.1106 / 0.1864
Long 4097/8192 | 0.0206 / 0.0721 / 0.0877 | 0.0199 / 0.0692 / 0.0692

<p class="isSelectedEnd"><span>The remaining issue is the non-production Hv48 T128 case: FlashInfer has five localized state-tail violations, compared with two for Triton. These five violations have been confirmed to originate from the upstream FlashInfer kernel rather than the OpenInfer HKV patch.</span></p><p><span>Production dispatch, whether to retain the SM120 Triton path, artifact distribution, and release preparation are still pending maintainer review.</span></p>
</body>
</html>
